File: source-code/symbolic-AI/n_queens.py
# n_queens.py - Solve the 8-queens problem using Python + Swi-Prolog
#
# Demonstrates calling Prolog from Python via swiplserver. The Prolog
# program n_queens.pl uses the clpfd (Constraint Logic Programming over
# Finite Domains) library to find all 92 solutions for placing 8 queens
# on a chessboard so that no two queens threaten each other.
#
# Requirements: brew install swi-prolog && uv pip install swiplserver

from swiplserver import PrologMQI
from pprint import pprint

with PrologMQI() as mqi:
    with mqi.create_thread() as prolog_thread:
        # Load the constraint library and the n_queens rules
        prolog_thread.query("use_module(library(clpfd)).")
        prolog_thread.query("[n_queens].")

        # Solve: find all placements of 8 queens.
        # label(Qs) triggers the constraint solver to enumerate solutions.
        # Each solution is a list of row positions for queens in columns 1-8.
        result = prolog_thread.query("n_queens(8, Qs), label(Qs).")
        pprint(result)
        print(len(result))  # should print 92

File: source-code/symbolic-AI/n_queens.pl
%% n_queens.pl - N-Queens constraint satisfaction program
%%
%% Uses Swi-Prolog's clpfd (Constraint Logic Programming over Finite Domains)
%% library to solve the N-Queens problem. Given N, finds all ways to place N
%% queens on an NxN board such that no two queens share a row, column, or
%% diagonal.
%%
%% Adapted from the Swi-Prolog documentation:
%%   https://www.swi-prolog.org/pldoc/man?section=clpfd-n-queens
%%
%% Usage from Prolog REPL:
%%   ?- use_module(library(clpfd)).
%%   ?- [n_queens].
%%   ?- n_queens(8, Qs), label(Qs).

:- use_module(library(clpfd)).

%% n_queens(+N, -Qs): Qs is a list of N row positions (one per column).
n_queens(N, Qs) :-
    length(Qs, N),
    Qs ins 1..N,        % each queen's row is in the range 1..N
    safe_queens(Qs).

%% safe_queens(+Qs): all queens in the list are mutually non-attacking.
safe_queens([]).
safe_queens([Q|Qs]) :-
    safe_queens(Qs, Q, 1),  % check Q against every queen to its right
    safe_queens(Qs).         % recursively check remaining queens

%% safe_queens(+Qs, +Q0, +D0): queen Q0 does not attack any queen in Qs,
%% where D0 is the column distance between Q0 and the head of Qs.
safe_queens([], _, _).
safe_queens([Q|Qs], Q0, D0) :-
    Q0 #\= Q,               % not on the same row
    abs(Q0 - Q) #\= D0,     % not on the same diagonal
    D1 #= D0 + 1,
    safe_queens(Qs, Q0, D1).
